src/Functions/changeDate.cpp

```diff
@@ -0,0 +1,396 @@
+            max_year = 2299;
+        }
+
+        switch (Traits::component)
+        {
+            case Component::Year:
+                if (new_value < min_year)
+                    return min_date;
+                else if (new_value > max_year)
+                    return max_date;
+                year = static_cast<Int16>(new_value);
+                break;
+            case Component::Month:
+                if (new_value < 1 || new_value > 12)
+                    return min_date;
+                month = static_cast<UInt8>(new_value);
+                break;
+            case Component::Day:
+                if (new_value < 1 || new_value > 31)
+                    return min_date;
+                day = static_cast<UInt8>(new_value);
+                break;
+            case Component::Hour:
+                if (new_value < 0 || new_value > 23)
+                    return min_date;
+                hours = static_cast<UInt8>(new_value);
+                break;
+            case Component::Minute:
+                if (new_value < 0 || new_value > 59)
+                    return min_date;
+                minutes = static_cast<UInt8>(new_value);
+                break;
+            case Component::Second:
+                if (new_value < 0 || new_value > 59)
+                    return min_date;
+                seconds = static_cast<UInt8>(new_value);
+                break;
+        }
+
+        Int64 result;
+        if (isDateOrDate32(result_type))
+            result = date_lut.makeDayNum(year, month, day);
+        else if (isDateTime(result_type))
+            result = date_lut.makeDateTime(year, month, day, hours, minutes, seconds);
+        else
+            result = DecimalUtils::decimalFromComponents<DateTime64>(
+                date_lut.makeDateTime(year, month, day, hours, minutes, seconds),
+                static_cast<Int64>(fraction),
+                static_cast<UInt32>(scale));
+
+        if (result < min_date)
+            return min_date;
+
+        if (result > max_date)
+            return max_date;
+
+        return result;
+    }
+};
+
+
+struct ChangeYearTraits
+{
+    static constexpr auto name = "changeYear";
+    static constexpr auto component = Component::Year;
+};
+
+struct ChangeMonthTraits
+{
+    static constexpr auto name = "changeMonth";
+    static constexpr auto component = Component::Month;
+};
+
+struct ChangeDayTraits
+{
+    static constexpr auto name = "changeDay";
+    static constexpr auto component = Component::Day;
+};
+
+struct ChangeHourTraits
+{
+    static constexpr auto name = "changeHour";
+    static constexpr auto component = Component::Hour;
+};
+
+struct ChangeMinuteTraits
+{
+    static constexpr auto name = "changeMinute";
+    static constexpr auto component = Component::Minute;
+};
+
+struct ChangeSecondTraits
+{
+    static constexpr auto name = "changeSecond";
+    static constexpr auto component = Component::Second;
+};
+
+}
+
+REGISTER_FUNCTION(ChangeDate)
```

Comment:
Could you please add docs of the new functions to https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions ?

Ideally, fields `syntax`, `arguments`, `returned_value`, `examples` of `FunctionDocumentation` are also provided.

(I am asking because there is an ongoing effort to document undocumented functions https://github.com/ClickHouse/clickhouse-docs/issues/1833)